### ISSUE TITLE
BUG: (Python 3.11) Compiler cl can not compile programs 

### ISSUE THREAD

**Issue by mxrch** [scipy/scipy#16988]
### Describe your issue.

Hello !
While trying to install `scipy` on Python 3.11rc1 (Windows 11), I got the error "Compiler cl can not compile programs".

I searched about the error `Compiler cl can not compile programs` but I didn't find anything to fix this problem, some things about Visual Studio but I already have VS 2019..
My main concern is that I'm updating my own open-source projects to be compatible with Python 3.11 (candidate version in 1 month), and I'm afraid that `scipy 1.9.1` doesn't work out of the box on Windows.

In addition, I confirm having installed `meson` using pip, and it's in PATH.
I thought at first that it was a meson issue so I opened an issue in `meson` (https://github.com/mesonbuild/meson/issues/10795), but it appears I'm not the only one having issues with `scipy 1.9.1` : https://stackoverflow.com/questions/73628794/unable-to-install-scipy-with-mkl-using-meson.

Also tried to manually build it, but after having installed `pythran`, `gfortran`, `cython`, `pkg-config`, `cmake` and tried to install `openblas`, I'm having a `scipy\scipy\meson.build:141:0: ERROR: Dependency "OpenBLAS" not found, tried pkgconfig and cmake` error. But that's not the subject of this issue.

### Reproducing Code Example

```python
py -m pip install scipy --upgrade
```

### Error message

```shell
Collecting scipy
  Using cached scipy-1.9.1.tar.gz (42.0 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [47 lines of output]
      The Meson build system
      Version: 0.62.2
      Source dir: C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e
      Build dir: C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e\.mesonpy-x7d63hr0\build
      Build type: native build
      Project name: SciPy
      Project version: 1.9.1
      Activating VS 16.11.18

      ..\..\meson.build:1:0: ERROR: Compiler cl can not compile programs.

      A full log can be found at C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e\.mesonpy-x7d63hr0\build\meson-logs\meson-log.txt
      + meson setup --native-file=C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e\.mesonpy-native-file.ini -Ddebug=false -Doptimization=2 --prefix=C:\Users\thher\AppData\Local\Programs\Python\Python311 C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e C:\Users\thher\AppData\Local\Temp\pip-install-9tovb2ma\scipy_828fda3b17134154918e54876239f05e\.mesonpy-x7d63hr0\build
      Traceback (most recent call last):
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 363, in <module>
          main()
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 130, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 923, in get_requires_for_build_wheel
          with _project(config_settings) as project:
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\contextlib.py", line 137, in __enter__
          return next(self.gen)
                 ^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 902, in _project
          with Project.with_temp_working_dir(
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\contextlib.py", line 137, in __enter__
          return next(self.gen)
                 ^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 547, in with_temp_working_dir
          yield cls(source_dir, tmpdir, build_dir)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 463, in __init__
          self._configure(reconfigure=bool(build_dir) and not native_file_mismatch)
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 494, in _configure
          self._meson(
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 477, in _meson
          return self._proc('meson', *args)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\thher\AppData\Local\Temp\pip-build-env-wder1mlf\overlay\Lib\site-packages\mesonpy\__init__.py", line 472, in _proc
          subprocess.check_call(list(args))
        File "C:\Users\thher\AppData\Local\Programs\Python\Python311\Lib\subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['meson', 'setup', '--native-file=C:\\Users\\thher\\AppData\\Local\\Temp\\pip-install-9tovb2ma\\scipy_828fda3b17134154918e54876239f05e\\.mesonpy-native-file.ini', '-Ddebug=false', '-Doptimization=2', '--prefix=C:\\Users\\thher\\AppData\\Local\\Programs\\Python\\Python311', 'C:\\Users\\thher\\AppData\\Local\\Temp\\pip-install-9tovb2ma\\scipy_828fda3b17134154918e54876239f05e', 'C:\\Users\\thher\\AppData\\Local\\Temp\\pip-install-9tovb2ma\\scipy_828fda3b17134154918e54876239f05e\\.mesonpy-x7d63hr0\\build']' returned non-zero exit status 1.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

(No more interesting details are present in the pip log file.)
```

### SciPy/NumPy/Python version information

1.23.2 sys.version_info(major=3, minor=11, micro=0, releaselevel='candidate', serial=1) (scipy not installed)

**Comment by mxrch** [comment]
For a temporary workaround, the wheel artifact built by the action here : https://github.com/scipy/scipy/actions/runs/2908455453#artifacts (cp311-win) is working perfectely.
```powershell
PS C:\Users\thher\Downloads> py -m pip install .\scipy-1.10.0.dev0-cp311-cp311-win_amd64.whl
Processing c:\users\thher\downloads\scipy-1.10.0.dev0-cp311-cp311-win_amd64.whl
Requirement already satisfied: numpy>=1.19.5 in c:\users\thher\appdata\local\programs\python\python311\lib\site-packages (from scipy==1.10.0.dev0) (1.23.2)
Installing collected packages: scipy
Successfully installed scipy-1.10.0.dev0
```

**Comment by andyfaff** [comment]
You'll have issues building with Windows at the moment with meson at the moment if you don't compile with `mingw-w64/gfortran`. We're currently looking into building SciPy with meson on windows at the moment with compilers other than `mingw-w64`, but that's a work in progress.
In addition those artefacts will get uploaded as *nightly wheels*, but we haven't activated that just yet because the wheel builder script is still relatively new.

**Comment by mxrch** [comment]
Thanks for the quick response, good to hear it's in the way and I'm not doing something wrong ! 😄 
Does this issue can get updated and closed when 3.11 wheels will be available for everyone on pypi ?

**Comment by rgommers** [comment]
Well, we still want to make sure that MSVC + MKL works. I think I already have the fixes needed for that in gh-16957. I'll separate those out and get them merged, so you can try. The rest of that PR is adding CI so we're sure it is working; fiddling with download commands to get MKL installed is the time-consuming part there.

**Comment by rgommers** [comment]
I opened gh-16991. @mxrch that should get you further, and hopefully things should work fully now. I just don't have a Windows machine to test. Could you try?

**Comment by mxrch** [comment]
@rgommers the https://github.com/scipy/scipy/pull/16991 PR seems good for the MSVC & Intel Fortran, but I have the OpenBLAS error back now : 
```
PS C:\Users\thher> py -m pip uninstall scipy
Could not find platform dependent libraries <exec_prefix>
Found existing installation: scipy 1.10.0.dev0
Uninstalling scipy-1.10.0.dev0:
  Would remove:
    c:\users\thher\appdata\local\programs\python\python311\lib\site-packages\scipy-1.10.0.dev0-cp311-cp311-win_amd64.whl
    c:\users\thher\appdata\local\programs\python\python311\lib\site-packages\scipy-1.10.0.dev0.dist-info\*
    c:\users\thher\appdata\local\programs\python\python311\lib\site-packages\scipy.libs\.load-order-scipy-1.10.0.dev0
    c:\users\thher\appdata\local\programs\python\python311\lib\site-packages\scipy.libs\libopenblas-57db09cfe174768fb409a6bb5a530d4c.dll
    c:\users\thher\appdata\local\programs\python\python311\lib\site-packages\scipy\*
Proceed (Y/n)? y
  Successfully uninstalled scipy-1.10.0.dev0
PS C:\Users\thher> py -m pip install git+https://github.com/rgommers/scipy@msvc-ifort-fixes
Could not find platform dependent libraries <exec_prefix>
Collecting git+https://github.com/rgommers/scipy@msvc-ifort-fixes
  Cloning https://github.com/rgommers/scipy (to revision msvc-ifort-fixes) to c:\users\thher\appdata\local\temp\pip-req-build-uf7kawtg
  Running command git clone --filter=blob:none --quiet https://github.com/rgommers/scipy 'C:\Users\thher\AppData\Local\Temp\pip-req-build-uf7kawtg'
  Running command git checkout -b msvc-ifort-fixes --track origin/msvc-ifort-fixes
  Branch 'msvc-ifort-fixes' set up to track remote branch 'msvc-ifort-fixes' from 'origin'.
  Switched to a new branch 'msvc-ifort-fixes'
  Resolved https://github.com/rgommers/scipy to commit 9e79aa8b71417a9425affa3e8bd8eadbeb6424e9
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [77 lines of output]
      Could not find platform dependent libraries <exec_prefix>
      Could not find platform dependent libraries <exec_prefix>
      The Meson build system
      Version: 0.63.2
      Source dir: C:\Users\thher\AppData\Local\Temp\pip-req-build-uf7kawtg
      Build dir: C:\Users\thher\AppData\Local\Temp\pip-req-build-uf7kawtg\.mesonpy-a2fomhai\build
      Build type: native build
      Project name: SciPy
      Project version: 1.10.0.dev0
      C compiler for the host machine: gcc (gcc 12.1.0 "gcc (GCC) 12.1.0")
      C linker for the host machine: gcc ld.bfd 2.37
      C++ compiler for the host machine: c++ (gcc 12.1.0 "c++ (GCC) 12.1.0")
      C++ linker for the host machine: c++ ld.bfd 2.37
      Host machine cpu family: x86_64
      Host machine cpu: x86_64
      Compiler for C supports arguments -Wno-unused-but-set-variable: YES
      Compiler for C supports arguments -Wno-unused-function: YES
      Compiler for C supports arguments -Wno-conversion: YES
      Compiler for C supports arguments -Wno-misleading-indentation: YES
      Compiler for C supports arguments -Wno-incompatible-pointer-types: YES
      Library m found: YES
      Fortran compiler for the host machine: gfortran (gcc 12.1.0 "GNU Fortran (GCC) 12.1.0")
      Fortran linker for the host machine: gfortran ld.bfd 2.37
      Compiler for Fortran supports arguments -Wno-conversion: YES
      Program cython found: YES (C:\Users\thher\AppData\Local\Temp\pip-build-env-m19b8f0h\overlay\Scripts\cython.EXE)
      Program pythran found: YES (C:\Users\thher\AppData\Local\Temp\pip-build-env-m19b8f0h\overlay\Scripts\pythran.EXE)
      Program scipy/_build_utils/copyfiles.py found: YES (C:\Windows\py.exe C:\Users\thher\AppData\Local\Temp\pip-req-build-uf7kawtg\scipy/_build_utils/copyfiles.py)
      Program python3 found: YES (C:\Windows\py.exe)
      WARNING: You should add the boolean check kwarg to the run_command call.
               It currently defaults to false,
               but it will default to true in future releases of meson.
               See also: https://github.com/mesonbuild/meson/issues/9300
      Library npymath found: YES
      Library npyrandom found: YES
      Found pkg-config: C:\ProgramData\chocolatey\bin\pkg-config.EXE (0.28)
      Found CMake: C:\Program Files\CMake\bin\cmake.EXE (3.24.1)
      Run-time dependency openblas found: NO (tried pkgconfig and cmake)
      Run-time dependency openblas found: NO (tried pkgconfig and cmake)

      ..\..\scipy\meson.build:140:0: ERROR: Dependency "OpenBLAS" not found, tried pkgconfig and cmake
```
I don't know if it's intended and we need to install all these dependencies to install `scipy` through `pip` ?

**Comment by rgommers** [comment]
You cannot install any BLAS/LAPACK library through `pip`. Installing with MKL should work. Can you try this:
```bash
# Note: the $PWD/build-install bit should be translated to a Windows-specific flavor
$ meson setup build -Dblas=mkl_rt -Dlapack=mkl_rt --prefix=$PWD/build-install -Ddebug=false -Doptimization=2
$ python -m build --wheel -Cbuilddir=build
```
That should work I believe - it did for Spack. In case MKL cannot be found, you have to locate the directory in which `libmkl_rt.dll` is located and set the environment variable `PKG_CONFIG_PATH` to that directory.

Changes in `meson` and `meson-python` are still needed to simplify the above to:
```bash
$ python -m build -Dblas=mkl -Dlapack=mkl
```

**Comment by rgommers** [comment]
For the record, I tried with the MKL from conda-forge, and because it doesn't have pkg-config files shipped (known issue, hopefully Intel fixes that) I had to use CMake for dependency detection (note that it uses the upper-case name):
```
$ meson setup build -Dblas=MKL -Dlapack=MKL --prefix=$PWD/build-install
$ python dev.py build
```
That segfaults though, because the g77 ABI wrappers are missed with the upper-case MKL name. And after fixing that it still doesn't work, because it doesn't link with `-lmkl_rt` like `numpy.distutils` does but with:
```
{
      "name":"MKL",
      "version":"2022.1.0",
      "compile_args":[
         "-I/home/rgommers/anaconda3/envs/scipy-dev/include"
      ],
      "link_args":[
         "-ldl",
         "-lm",
         "-lpthread",
         "/home/rgommers/anaconda3/envs/scipy-dev/lib/libiomp5.so",
         "/home/rgommers/anaconda3/envs/scipy-dev/lib/libmkl_core.so",
         "/home/rgommers/anaconda3/envs/scipy-dev/lib/libmkl_gf_ilp64.so",
         "/home/rgommers/anaconda3/envs/scipy-dev/lib/libmkl_intel_thread.so"
      ]
   },
```
It wants to use the ILP64 interface instead of our defaulting to 32-bit.

tl;dr some work left to do here.

**Comment by rgommers** [comment]
MKL has issues, tracking issue is gh-16200. Regarding this issue, MSVC should work after gh-16991 is merged. I expect it to do the right thing when used with Intel Fortran and OpenBLAS.

**Comment by rgommers** [comment]
I'll close this issue, there isn't anything actionable left to do here. The fixes linked above were merged, and there are tracking issues for the rest.
